libs/community/langchain_community/chat_models/llamacpp.py

```diff
@@ -0,0 +1,579 @@
+from operator import itemgetter
+from pathlib import Path
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Iterator,
+    List,
+    Optional,
+    Sequence,
+    Type,
+    Union,
+    cast,
+)
+
+from langchain_core.callbacks import CallbackManagerForLLMRun
+from langchain_core.language_models import LanguageModelInput
+from langchain_core.language_models.chat_models import (
+    BaseChatModel,
+    generate_from_stream,
+)
+from langchain_core.messages import (
+    AIMessageChunk,
+    BaseMessage,
+)
+from langchain_core.output_parsers.base import OutputParserLike
+from langchain_core.output_parsers.openai_tools import (
+    JsonOutputKeyToolsParser,
+    PydanticToolsParser,
+)
+from langchain_core.outputs import ChatGeneration, ChatGenerationChunk, ChatResult
+from langchain_core.pydantic_v1 import BaseModel, Field, root_validator
+from langchain_core.runnables import Runnable, RunnableMap, RunnablePassthrough
+from langchain_core.tools import BaseTool
+from langchain_core.utils.function_calling import convert_to_openai_tool
+
+from langchain_openai.chat_models.base import (
+    _convert_delta_to_message_chunk,
+    _convert_dict_to_message,
+    _convert_message_to_dict
+)
```

Comment:
shouldn't be relying on the langchain-openai package since its not a langchain-community dependency

Reply:
Agreed

Reply:
Ya. FWIW, I manually copied this function over to test on my branch.

That seems reasonable, and mirrors what a number of the other other chat models do.